### ISSUE TITLE
How does the It.IsIn<Exception> call work for verifying calls to the logger?

### ISSUE THREAD

**Issue by macfarmw** [devlooped/moq#1229]
I'm using Moq to verify the logger gets called with Warning level if 2 specific exception types are thrown. The exception instances are being passed to the logger.LogWarning method. The Mock verifies the the log level successfully, but I can't figure out how to verify the exception type. Here is my attempt:

```csharp
logger.Verify(x => x.Log(
    It.Is<LogLevel>(l => l == LogLevel.Warning),
    It.IsAny<EventId>(),
    It.IsAny<It.IsAnyType>(),
    It.IsIn<Exception>(new TaskCanceledException(), new OperationCanceledException()),
    (Func<It.IsAnyType, Exception, string>)It.IsAny<object>()));
```

This compiles, but when the class catches these exceptions and logs the warning Verify throws the exception:

`Expected invocation on the mock at least once, but was never performed: x => x.Log<It.IsAnyType>(It.Is<LogLevel>(l => (int)l == 3), It.IsAny<EventId>(), It.IsAny<It.IsAnyType>(), It.IsIn<Exception>([System.Threading.Tasks.TaskCanceledException: A task was canceled., System.OperationCanceledException: The operation was canceled.]), (Func<It.IsAnyType, Exception, string>)It.IsAny<object>())`

What is the correct syntax for setting up the list of expected exception types?

**Comment by stakx** [comment]
It seems unlikely (actually, almost impossible) that any specific call would throw _exactly_ the exception instance that you're `new`-ing up in your test code. (`Exception` and thus all derived classes are reference types, so what gets compared by default is their identity, not their types/values.)

You probably meant to do something like `It.Is(ex => ex is FooException || ex is BarException)` (or using some other form of pattern matching)?

**Comment by macfarmw** [comment]
Hi Dominique,

Yes, that's exactly what I meant to do 😄! 

After I applied your suggestion I noticed that the verify was still not matching on the exception types. I took a closer look at the calling code and found the the call to `LogWarning` needed to be changed.

I changed
`Logger.LogWarning("Integration {IntegrationName} {@Exception}", Integration.IntegrationName, tcex);`

to
`Logger.LogWarning(tcex, "Integration {IntegrationName}", Integration.IntegrationName);`

By using the overload that takes the Exception instance first the mock Verify can see the exception and match successfully. Done!

Thanks for your help and all the great work on Moq.
